src/components/Hooks/fetchSortedContacts.js

```diff
@@ -21,30 +23,23 @@ const fetchSortedContacts = hasServiceBeenLaunched => {
     contactsWithNoIndexes
   ] = useFetchContacts(contactsWithoutIndexes)
 
-  const [hasAlreadySortedContacts, setHasAlreadySortedContacts] = useState(
-    false
-  )
-  const [sortedContacts, setSortedContacts] = useState([])
-
   const fetchesAreFinished =
     isContactsWithIndexesFinished && isContactsWithNoIndexesFinished
 
-  if (fetchesAreFinished && !hasAlreadySortedContacts) {
-    // hasServiceBeenLaunched will be deprecated when
-    // all instances have been migrated.
-    // Therefore, the false condition will no longer be needed and can be removed.
-    const reworkedContacts = hasServiceBeenLaunched
-      ? contactsWithIndexes.concat(contactsWithNoIndexes)
-      : harmonizeAndSortByFamilyNameGivenNameEmailCozyUrl(
+  useEffect(
+    () => {
+      if (fetchesAreFinished) {
+        const reworkedContacts = reworkContacts(
+          hasServiceBeenLaunched,
           contactsWithIndexes,
           contactsWithNoIndexes
         )
+        setSortedContacts(reworkedContacts)
+      }
+    },
+    [JSON.stringify(contactsWithIndexes), JSON.stringify(contactsWithNoIndexes)]
```

Comment:
Are you doing that because you want a deep equality check? https://github.com/kentcdodds/use-deep-compare-effect should be better. 

Even with the use of this deepCompareEffect, it's not confortable with this kind of check. No idea for now to how to do it better. 

Maybe @y-lohse or @ptbrowne have a few 💡

Reply:
Yes it's because as it's said in use-deep-compare-effect `even if none of the properties changed, your effect will get called anyway` but also because React can't compare array or object unless they are stringified.

Reply:
https://medium.com/better-programming/why-you-shouldnt-use-json-stringify-to-compare-objects-in-javascript-c9a16b7331e 🎁

Reply:
Damned !

Reply:
I'm confused, why do we want to switch to a `useEffect`?

Reply:
@Crash--  For what I saw https://twitter.com/dan_abramov/status/1104414272753487872 

`useEffect` is used to simply do something (or not) depends on a data comparison. `setSortedContacts` is triggered only if queries result change, to prevent infinite loop

Reply:
@Crash-- I'm thinking about JS objects and stringify. Here it's an array of fetched data, do you think we'll encouter the same problem ?

Reply:
> but also because React can't compare array, or object, or array of objects unless they are stringified.

As far as the data is coming from cozy-client (I see that it is coming from a useQuery call, thus it comes from the cozy-client store), it is safe to assume that it has been treated as an immutable object : inside cozy-client, we do not mutate objects in place, precisely so that you can rely on reference equality checks downstream.

To stringify the thousands of contacts to check if we need to render is very inefficient :/

I see two potential solutions:

- `useMemo` can be used here.
- Using reselect selectors so that there is a sorted "view" available for multiple objects to rely on.

Reply:
Another idea, sounds too simple to work so I'm probably missing something: don't use `useEffect`, and don't store the `sortedContacts` in a `useState`, simply return them. 

As far as I understand, `useEffect` is only there to prevent the infinite loop triggered by `setSortedContacts`. But I don't see why we keep the sorted contacts around anyway, they are computed dynamically every time the hook runs..?

Reply:
This will cause performance issue because everything is triggered/re-rendered after clicking on a contact in the list

Reply:
Why that? What is the root cause of this new render?

Reply:
Hooks calls rerender the component not matter what. By default the useQuery hook make 3 renders to get data, even more if we need to fetchMore, as well as its "parents". So I think there is too much useless render.
I found https://github.com/facebook/react/issues/14476#issuecomment-471199055 and will explore `useMemo` way

Reply:
The 3 renders that you talk about are the initial renders, not why we need a full re-render of the list after a click on the contact.

Reply:
The full list is re-rendered here because the contact modal is shown at the list level. If it was done at row level, only the row would be re-rendered.

Reply:
I tried but the behavior is the same : `ContentWrapper` and so the list, are re-rendered